server/neptune/workflows/internal/deploy/revision/queue/revision_processor.go

```diff
@@ -0,0 +1,114 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config/logger"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/deploy/terraform"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/github"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/root"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+)
+
+type githubActivities interface {
+	CompareCommit(ctx context.Context, request activities.CompareCommitRequest) (activities.CompareCommitResponse, error)
+	UpdateCheckRun(ctx context.Context, request activities.UpdateCheckRunRequest) (activities.UpdateCheckRunResponse, error)
+}
+
+type RevisionProcessor struct {
+	Activities githubActivities
+}
+
+const (
+	DirectionBehindSummary   = "This revision is behind the current revision and will not be deployed.  If this is intentional, revert the default branch to this revision to trigger a new deployment."
+	UpdateCheckRunRetryCount = 5
+
+	ForceApplySummary = "The current deployment has diverged from the default branch, so we have locked the root. This is most likely the result of this PR performing a deployment. To override that lock and allow the main branch to perform new deployments, select the Unlock button."
+)
+
+func (p *RevisionProcessor) Process(ctx workflow.Context, requestedDeployment terraform.DeploymentInfo, latestDeployment *root.DeploymentInfo) error {
+	commitDirection, err := p.getDeployRequestCommitDirection(ctx, requestedDeployment, latestDeployment)
+	if err != nil {
+		return err
+	}
+
+	// TODO: remove log? it might be noisy
+	logger.Info(ctx, fmt.Sprintf("relationship of deployed to requested revision: %s", commitDirection),
+		"deployed-revision", latestDeployment.GetRevision(),
+		"requested-revision", requestedDeployment.Revision)
+
+	switch commitDirection {
+	case activities.DirectionBehind:
+		p.updateCheckRun(ctx, requestedDeployment, github.CheckRunFailure, DirectionBehindSummary)
+		return fmt.Errorf("requested revision %s is behind current one %s", requestedDeployment.Revision, latestDeployment.GetRevision())
```

Comment:
What is the best practice on returning an error in go? The calling function wraps the returned error in `failed to process revision`. For commits that are behind the main branch, we technically processed the revision as expected (`by updating the checkrun`) right? Not sure if there's a better way to handle this but just wanted to point it out!

Reply:
I guess this may not be best practice, but the main reason I made this into an error was to not overcomplicate the check in the calling function (if Process() returns an error, then skip the revision).